src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -597,4 +597,138 @@ ValidatorList::calculateMinimumQuorum (
+    std::shuffle (indexes.begin(), indexes.end(), crypto_prng());
+
+    for (auto const& index : indexes)
+    {
+        auto const& val = std::next (keyListings_.begin(), index);
+
+        if (validatorManifests_.revoked (val->first))
+            continue;
+
+        if (val->first == localPubKey_)
+        {
+            localKeyListed = val->second > 1;
+            rankedKeys.insert (
+                std::pair<std::size_t,PublicKey>(
+                    std::numeric_limits<std::size_t>::max(), localPubKey_));
+        }
+        // If the total number of validators is too small, or
+        // no validations are being received, use all validators.
+        // Otherwise, do not use validators whose validations aren't
+        // being received.
+        else if (
+            keyListings_.size() < MINIMUM_RESIZEABLE_UNL ||
+            seenValidators.empty() ||
+            seenValidators.find(calcNodeID(val->first)) != seenValidators.end())
+        {
+            rankedKeys.insert (
+                std::pair<std::size_t,PublicKey>(val->second, val->first));
+        }
+    }
+
+    // This minimum quorum guarantees safe overlap with the trusted sets of
+    // other nodes using the same set of published lists.
+    std::size_t quorum = calculateMinimumQuorum (keyListings_.size(),
+        localPubKey_.size() && !localKeyListed);
+
+    JLOG (j_.debug()) <<
+        rankedKeys.size() << "  of " << keyListings_.size() <<
+        " listed validators eligible for inclusion in the trusted set";
+
+    auto size = rankedKeys.size();
+
+    // Require 80% quorum if there are lots of validators.
+    if (rankedKeys.size() > BYZANTINE_THRESHOLD)
+    {
+        // Use all eligible keys if there is only one trusted list
+        if (publisherLists_.size() == 1 ||
+                keyListings_.size() < MINIMUM_RESIZEABLE_UNL)
+        {
+            // Try to raise the quorum to at least 80% of the trusted set
+            quorum = std::max(quorum, size - size / 5);
+        }
+        else
+        {
+            // Reduce the trusted set size so that the quorum represents
+            // at least 80%
+            size = quorum * 1.25;
+        }
+    }
+
+    if (minimumQuorum_ && seenValidators.size() < quorum)
+    {
+        quorum = *minimumQuorum_;
+        JLOG (j_.warn())
+            << "Using unsafe quorum of "
+            << quorum_
+            << " as specified in the command line";
+    }
+
+    // Do not use achievable quorum until lists from all configured
+    // publishers are available
+    else if (! allListsAvailable)
+        quorum = std::numeric_limits<std::size_t>::max();
+
+    TrustChanges trustChanges;
+    hash_set<PublicKey> newTrustedKeys;
+    for (auto const& val : boost::adaptors::reverse(rankedKeys))
+    {
+        if (size <= newTrustedKeys.size())
+            break;
+        newTrustedKeys.insert(val.second);
+
+        if (trustedKeys_.count(val.second) == 0)
```

Comment:
You could change this to
```
if (trustedKeys_.erase(val.second) == 0)
```
and then you could remove the `if(newTrustedKeys.count(k) == 0)` check in the loop  below